feat(ctrl): implements MeshFederation reconciliation

bartoszmajsak opened the pull request:
This commit introduces the initial implementation of the MeshFederation
controller. The controller is responsible for:

- Managing the MeshFederation server lifecycle (#152)
- Configuring MeshFederation resources, including:
  - Ingress `Gateway`
  - `PeerAuthentication`
  - `EnvoyFilter` (for OpenShift Router)
  - `Route`s (for OpenShift Router)
- Watching Kubernetes `Services` to:
  - Push SotW updates to all connected peers (#153)
  - Update `MeshFederation` cluster configuration
  
Support both label selectors and expressions (#52 #143) as export rules.

Basic EnvTest tests are included to verify the setup (happy paths for both 
`istio` and `openshift-router` ingress types).

Introducing `MeshFederation.Status.ExportedServices` to keep track of 
exported services in each reconcile.

### Current limitations / random thoughts 

There are loose ends I identified while implementing it. Most are in `TODO`s, 
but I thought keeping them in the PR description helps with getting initial context.

- not tested for multiple `MeshFederation` instances
    - we currently create one FDS server per federation, but we have a single port for 
      all which is hardcoded (`:15080`), some ideas:
       - share server across federations
          - ❓ is that a bottleneck? 
          - we are lacking a way to identify to which federation a remote belongs to, 
            which makes it impossible to e.g. push SotW to only relevant peers on a given
            `MeshFederation` reconcile. Perhaps expanding FDS protocol can solve it.
       - alternatively we can make the port customizable, but that would mean configuring 
         all remotes in a single federation instead of relying on the defaults
    - defaulting root ns (`istio-system`) can cause issues when we create mesh-wide resources)
- not ensuring one-per-ns `MeshFederation`
    -  should be done through admission webhook if that becomes a requirement


Fixes #152
Fixes #52
Fixes #143 
Fixes #153

bartoszmajsak committed:
feat(ctrl): implement MeshFederation reconciliation

This commit introduces the initial implementation of the MeshFederation
controller. The controller is responsible for:

- Managing the MeshFederation server lifecycle (#152)
  - removes a need for a channel to trigger push - pushing directly
    instead
- Configuring MeshFederation resources, including:
  - IngressGateway
  - PeerAuthentication
  - EnvoyFilter (for OpenShift Router)
  - Routes (for OpenShift Router)
- Watching Kubernetes services to:
  - Push SotW updates to all connected peers (#153)
  - Update MeshFederation cluster configuration
  - Support both label selectors and expressions (#52 #143)

Basic EnvTest tests are included to verify the setup.

Fixes #152 #52 #143 #153

bartoszmajsak commented:
This PR also includes #167 - hence the size.

eoinfennessy commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,79 +16,272 @@ package meshfederation
 		_, errStatusUpdate := controller.RetryStatusUpdate(ctx, r.Client, meshFederation, func(saved *v1alpha1.MeshFederation) {
 			for _, condition := range conditions {
 				machinerymeta.SetStatusCondition(&saved.Status.Conditions, condition)
+				saved.Status.ExportedServices = meshFederation.Status.ExportedServices
```
This can be moved out of the loop I think.

bartoszmajsak replied:
Of course! Thank you.

78723aa

eoinfennessy reviewed (commented):
Initial pass with a few comments/suggestions. I'll continue the review later.

bartoszmajsak committed:
chore: makes finalizerHandler part of reconcile struct

bartoszmajsak commented on internal/pkg/discovery/grpc_server.go:
```diff
@@ -0,0 +1,102 @@
+func (s *Server) Start(ctx context.Context) bool {
+	if !s.IsRunning() {
+		go func() {
+			_ = s.Run(ctx)
```
Thanks for the thorough review. I think I lost the `// TODO` in the rush this morning. Wouldn't simply restarting on error be enough? What scenario would a reconcile help with?

bartoszmajsak replied:
I really appreciate the thorough review. I missed the `// TODO` in the rush this morning. Wouldn't simply restarting on error be enough? What scenario would a reconcile help with?

bartoszmajsak replied:
In 8655a60 I renamed the func to `StartOnce` and added simple retry logic with error logging.  Thanks for this suggestion!

eoinfennessy replied:
> I was thinking simple restart on error would be enough here? What scenario would a reconcile help with?

I think you are right. In the future we could maybe think about longer lived server errors and how the MeshFederation status could reflect them.

bartoszmajsak commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,35 +16,238 @@ package meshfederation
+	// TODO paginate? options?
+	// TODO handle multiple matching rules (as one is AND-ed, not OR-ed)
+	if errSvcList := r.Client.List(ctx, exportedServices, client.MatchingLabelsSelector{Selector: exportSelector}); errSvcList != nil {
+		return reconcile.Result{RequeueAfter: 2 * time.Second}, errSvcList
```
Why here? Not sure anymore - probably `List` is not the best example and this might not be needed here :)

My overall goal was to reduce API Server calls, hoping intermittent issues would resolve on their own. I also intended to define it as a const and reuse it elsewhere.

Why 2? No empirical evidence - just a power of 2, so it must be good! :)

bartoszmajsak replied:
I will go over the code and adjust requeues where it might help.

bartoszmajsak replied:
Probably `List` is not the best example.

My general intention was to avoid excessive calls to the API Server with the hopes that some intermittent issues can resolve in the meantime. I wanted to at least make it a `const` and put it in other relevant places.

As to why exactly 2? Do not have any empirical evidence, but it's a power of 2 so must be good! :)

I will review the code and apply requeue delays where they make sense,

eoinfennessy replied:
I was just curious more than anything! I wonder is there an established best practice/recommendation from controller-runtime folks?

bartoszmajsak committed:
chore: moves slice assignment outside of conditions loop

bartoszmajsak commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,35 +16,238 @@ package meshfederation
+
+	original := meshFederation.DeepCopy()
+
+	exportSelector, err := metav1.LabelSelectorAsSelector(meshFederation.Spec.ExportRules.ServiceSelectors)
+	if err != nil {
+		return reconcile.Result{RequeueAfter: 2 * time.Second}, fmt.Errorf("invalid label selector: %v", err)
+	}
```
Thanks! #173 so we do not forget.

bartoszmajsak replied:
Reworked to only log error for now. 4cb86dc

bartoszmajsak commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,35 +16,238 @@ package meshfederation
+		selector: exportSelector,
+	})
+
+	finalizerHandler := finalizer.NewHandler(r.Client, "federation.openshift-service-mesh.io/mesh-federation")
```
Good point, totally unnecessary. Fixed in 47d00a5

bartoszmajsak committed:
chore: logs error on selector error, does not reconcile

bartoszmajsak committed:
chore: corrects comment

bartoszmajsak committed:
chore: removes reduntant impl assertion

bartoszmajsak committed:
fix: weird auto-merge

bartoszmajsak committed:
fix: do not requeue on push-related errors

There is no point in stopping reconcile - we want to continue with bringing the cluster to the desired state.

We should:

- report on status
- raise event

bartoszmajsak committed:
chore: minor todos

bartoszmajsak committed:
chore: rename

bartoszmajsak committed:
chore: narrows service watch with generation or label change predicates

bartoszmajsak committed:
chore: renames to server.StartOnce

adds retry logic with delay

bartoszmajsak commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,35 +16,238 @@ package meshfederation
+
+	// TODO: figure out preferred approach to deal with conditions (metav1 vs conditionsv1 from Openshift)
+	// TODO: wrap conditions handling in a pkg/funcs representing domain-oriented conditions
+	justStarted := server.Start(ctx)
+	if justStarted {
+		_ = machinerymeta.SetStatusCondition(&meshFederation.Status.Conditions, metav1.Condition{
+			Type:    "FDSServerRunning",
+			Status:  "True",
+			Reason:  "FDSServerStarted",
+			Message: "FDS Server has started",
+		})
+	}
```
I removed this condition b/c we used it initially for tests. I think we should connect the client instead to have a way of verifying what has actually been passed. We can get back to discussing what conditions we need later.

In 8655a60 I renamed the func to `StartOnce` and added simple retry logic with error logging.  Thanks for this suggestion!

bartoszmajsak replied:
Renamed to `StartOnce` in https://github.com/openshift-service-mesh/federation/commit/8655a605cfda20ba807c82256fc929d3fe72e1b7

bartoszmajsak commented:
Lets don't `ok-to-merge` yet, ideally we should combine it with #168, as there were API changes among other things.

bartoszmajsak commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,79 +16,238 @@ package meshfederation
+		Owns(&routev1.Route{}).
+		Watches(
+			// TODO(design): initial reconcile will trigger a lot of requests - one for each service. This can become expensive.
+			&corev1.Service{},
```
## FOLLOW-UP 
We don't need to watch the `corev1.Service` object, we can limit it only to what we are interested in at this point - labels. For this purpose watch on [`PartialObjectMetadata`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1beta1#PartialObjectMetadata) is sufficient. This will also reduce the payload send from k8s api-server.

@eoinfennessy I think you can also think about it for the zones as you rely on namespace only IIRC. Thing I don't know yet is if that's the same cache as regular objects.

eoinfennessy replied:
Thank you! I actually wrote a TODO for this very thing! :) https://github.com/openshift-service-mesh/istio-zones/blob/main/internal/controller/zone_controller.go#L256-L257

bartoszmajsak replied:
Yeah I saw that the other day and woke up with `PartialObjectMetadata` thing in my head so I just dumped it here before first coffee :)

eoinfennessy commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,79 +16,238 @@ package meshfederation
+	return accResult, errors.Join(errs...)
+}
+
+func (r *Reconciler) handleServiceToExport(ctx context.Context, object client.Object) []reconcile.Request {
```
Nit: I think the name could be a little clearer. Something like this:

```suggestion
func (r *Reconciler) mapServiceToReconcileRequests(ctx context.Context, object client.Object) []reconcile.Request {
```

eoinfennessy commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,79 +16,238 @@ package meshfederation
+	meshFederations := &v1alpha1.MeshFederationList{}
+	// TODO paginate? options?
+	if errList := r.Client.List(ctx, meshFederations); errList != nil {
+		logger.Error(errList, "failed mapping Service to MeshFederations", "service", object.GetName()+"/"+object.GetNamespace())
```
```suggestion
		logger.Error(errList, "failed to list MeshFederations when mapping Service to reconcile requests", "service", object.GetName()+"/"+object.GetNamespace())
```

eoinfennessy commented on internal/controller/meshfederation/reconcilers.go:
```diff
@@ -0,0 +1,293 @@
+		}
+	}
+
+	createGateway := func() objReconciler[*v1alpha3.Gateway, istionetv1alpha3.Gateway] {
+		result := &v1alpha3.Gateway{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "federation-ingress-gateway",
+				Namespace: meshFederation.Spec.ControlPlaneNamespace,
+				Labels:    map[string]string{"federation.openshift-service-mesh.io/peer": "todo"},
+			},
+		}
+
+		result.Spec = desiredSpec()
+
+		return objReconciler[*v1alpha3.Gateway, istionetv1alpha3.Gateway]{
+			Obj: result,
+			DesiredSpec: func() istionetv1alpha3.Gateway {
+				return desiredSpec()
+			},
+		}
+	}
```
Nit

```suggestion
	createGateway := func() objReconciler[*v1alpha3.Gateway, istionetv1alpha3.Gateway] {
		result := &v1alpha3.Gateway{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "federation-ingress-gateway",
				Namespace: meshFederation.Spec.ControlPlaneNamespace,
				Labels:    map[string]string{"federation.openshift-service-mesh.io/peer": "todo"},
			},
			Spec: desiredSpec(),
		}

		return objReconciler[*v1alpha3.Gateway, istionetv1alpha3.Gateway]{
			Obj: result,
			DesiredSpec: func() istionetv1alpha3.Gateway {
				return desiredSpec()
			},
		}
	}
```

eoinfennessy commented on internal/controller/meshfederation/reconcilers.go:
```diff
@@ -0,0 +1,293 @@
+		}
+	}
+
+	envoyFilter := func(svcName, svcNamespace string, port int32) objReconciler[*v1alpha3.EnvoyFilter, istionetv1alpha3.EnvoyFilter] {
+		result := &v1alpha3.EnvoyFilter{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("sni-%s-%s-%d", svcName, svcNamespace, port),
+				Namespace: meshFederation.Spec.ControlPlaneNamespace,
+				Labels: map[string]string{
+					"federation.openshift-service-mesh.io/peer": "todo",
+				},
+			},
+		}
+
+		result.Spec = desiredSpec(svcName, svcNamespace, port)
+
+		return objReconciler[*v1alpha3.EnvoyFilter, istionetv1alpha3.EnvoyFilter]{
+			Obj: result,
+			DesiredSpec: func() istionetv1alpha3.EnvoyFilter {
+				return desiredSpec(svcName, svcNamespace, port)
+			},
+		}
+	}
```
Nit

```suggestion
	envoyFilter := func(svcName, svcNamespace string, port int32) objReconciler[*v1alpha3.EnvoyFilter, istionetv1alpha3.EnvoyFilter] {
		result := &v1alpha3.EnvoyFilter{
			ObjectMeta: metav1.ObjectMeta{
				Name:      fmt.Sprintf("sni-%s-%s-%d", svcName, svcNamespace, port),
				Namespace: meshFederation.Spec.ControlPlaneNamespace,
				Labels: map[string]string{
					"federation.openshift-service-mesh.io/peer": "todo",
				},
			},
			Spec: desiredSpec(svcName, svcNamespace, port),
		}

		return objReconciler[*v1alpha3.EnvoyFilter, istionetv1alpha3.EnvoyFilter]{
			Obj: result,
			DesiredSpec: func() istionetv1alpha3.EnvoyFilter {
				return desiredSpec(svcName, svcNamespace, port)
			},
		}
	}
```

eoinfennessy commented on internal/controller/meshfederation/reconcilers.go:
```diff
@@ -0,0 +1,293 @@
+		}
+	}
+
+	createRoute := func(svcName, svcNamespace string, port int32) objReconciler[*routev1.Route, routev1.RouteSpec] {
+		result := &routev1.Route{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s-%d-to-federation-ingress-gateway", svcName, svcNamespace, port),
+				Namespace: meshFederation.Spec.ControlPlaneNamespace,
+				Labels:    map[string]string{"federation.openshift-service-mesh.io/peer": "todo"},
+			},
+		}
+
+		result.Spec = desiredSpec(svcName, svcNamespace, port)
+
+		return objReconciler[*routev1.Route, routev1.RouteSpec]{
+			Obj: result,
+			DesiredSpec: func() routev1.RouteSpec {
+				return desiredSpec(svcName, svcNamespace, port)
+			},
+		}
+	}
```
Nit
```suggestion
	createRoute := func(svcName, svcNamespace string, port int32) objReconciler[*routev1.Route, routev1.RouteSpec] {
		result := &routev1.Route{
			ObjectMeta: metav1.ObjectMeta{
				Name:      fmt.Sprintf("%s-%s-%d-to-federation-ingress-gateway", svcName, svcNamespace, port),
				Namespace: meshFederation.Spec.ControlPlaneNamespace,
				Labels:    map[string]string{"federation.openshift-service-mesh.io/peer": "todo"},
			},
			Spec: desiredSpec(svcName, svcNamespace, port),
		}

		return objReconciler[*routev1.Route, routev1.RouteSpec]{
			Obj: result,
			DesiredSpec: func() routev1.RouteSpec {
				return desiredSpec(svcName, svcNamespace, port)
			},
		}
	}
```

eoinfennessy commented on internal/pkg/discovery/adss_handler.go:
```diff
@@ -0,0 +1,198 @@
+		return nil
+	}
+
+	resources := pushRequest.Resources
+	if resources == nil {
+		var err error
+		resources, err = adss.generateResources(pushRequest.TypeUrl)
+		if err != nil {
+			return err
+		}
+	}
```
Readability suggestion avoiding variable reassignment:

```suggestion
	var resources []*anypb.Any
	if pushRequest.Resources == nil {
		var err error
		resources, err = adss.generateResources(pushRequest.TypeUrl)
		if err != nil {
			return err
		}
	} else {
		resources = pushRequest.Resources
	}
```

eoinfennessy reviewed (approved):
Looks great! I just added a few small nits/suggestions.

eoinfennessy commented on internal/controller/meshfederation/meshfederation_controller.go:
```diff
@@ -16,79 +16,238 @@ package meshfederation
-		// TODO: patch and call only when actually conditionsChanged
+		// TODO: patch and call only when conditions were changed
 		_, errStatusUpdate := controller.RetryStatusUpdate(ctx, r.Client, meshFederation, func(saved *v1alpha1.MeshFederation) {
+			// TODO(design): do we want to keep all the services? seems convenient, but shouldn't we be concerned about the size?
+			saved.Status.ExportedServices = meshFederation.Status.ExportedServices
```
## FOLLOW-UP

@jewertow @bartoszmajsak

It may be possible to minimise the number of full SotW pushes by comparing the old list of services to the new list of services and only pushing if the list has changed. This could be achieved by storing a hash of the services in the MeshFederation's status (`status.lastPushHash: 1234abcd`) and only pushing if this hash changes.
